### PR TITLE
[WIP] Diffs for hydration errors

### PR DESCRIPTION
**Pull request by gaearon**
- [ ] Errors
  - [x] Extra server element
  - [x] Extra server text
  - [x] Extra client element
  - [x] Extra client text
  - [ ] Extra server attributes
  - [ ] Prop mismatch
- [ ] Formatting
  - [x] Parent tag and attributes
  - [x] Sibling tag/text and attributes
  - [x] Child tag/text and attributes
  - [x] Ellipsis to indicate more siblings before or after
- [ ] Special cases
  - [ ] Document/fragment parent
  - [x] Long attributes
  - [x] Too many attributes
  - [ ] Quotes in attributes
  - [ ] Long text
  - [x] Skip over non-element/text siblings
- [ ] Other stuff
  - [ ] Flow

### CODE REVIEW

File: packages/react-dom/src/client/ReactDOMComponent.js [hunk 1]
```diff
@@ -1209,9 +1312,10 @@ export function warnForDeletedHydratableElement(
     }
     didWarnInvalidHydration = true;
     console.error(
-      'Did not expect server HTML to contain a <%s> in <%s>.',
-      child.nodeName.toLowerCase(),
-      parentNode.nodeName.toLowerCase(),
+      'The content rendered by the server and the client did not match ' +
```
Comment (sompylasar) [inline packages/react-dom/src/client/ReactDOMComponent.js:1315]: The explanation in detail sounds great, but I feel there are too many words that are the same across multiple messages until the very essence of this message which is "the server has rendered an extra text node". Could be easier and quicker to grasp the gist if the varying part of the message was at the beginning, and the repetitive explanation at the end.

Comment (gaearon) [reply packages/react-dom/src/client/ReactDOMComponent.js:1315]: Yeah good point. I'll take another pass to clean them up after all of them are covered.

**Comment by himself65** [comment]
Any update for this PR?

**Comment by gaearon** [comment]
No, I'm not actively working on this so there won't be any update.

**Comment by gaearon** [comment]
Someone motivated could pick this up though!

**Comment by himself65** [comment]
> Someone motivated could pick this up though!

I will try to continue when I'm avaliable!

**Comment by gaearon** [comment]
That would be great! The first step would probably be to get this rebased and/or rewrite it based on the tests I added and the logic I've included. Then you can look at the remaining TODOs.

File: packages/react-dom/src/client/ReactDOMComponent.js [hunk 2]
```diff
@@ -1200,6 +1200,109 @@ export function diffHydratedText(
...
+    str += '>\n' + formattedChildren;
+    str += '\n' + indentation;
+    str += '</' + element.nodeName.toLowerCase() + '>';
+  }
+  return str;
+}
+
+function formatNode(node, indentation) {
+  switch (node.nodeType) {
+    // TODO: use constants
```
Comment (himself65) [inline packages/react-dom/src/client/ReactDOMComponent.js:1290]: What does `use constants` mean here? Isn't here already constants?

**Comment by sebmarkbage** [comment]
Going with a more invasive refactor and different approach in #28512.

**Pull request closed without merging**